packages/server/lib/plugins/child/run_require_async_child.js

```diff
@@ -92,18 +104,36 @@ function run (ipc, configFile, projectRoot) {
 
         areSetupNodeEventsLoaded = true
         if (testingType === 'component') {
-          if (!isValidSetupNodeEvents(result.component?.setupNodeEvents)) {
+          if (!isValidSetupNodeEvents(result.setupNodeEvents) || !isValidDevServer((result.component || {}))) {
             return
           }
 
           runPlugins.runSetupNodeEvents(options, (on, config) => {
-            if (result.component?.devServer) {
-              on('dev-server:start', (options) => result.component.devServer(options, result.component?.devServerConfig))
+            const setupNodeEvents = result.component?.setupNodeEvents ?? ((on, config) => {})
+
+            const { devServer } = result.component
+
+            if (typeof devServer.devServer === 'function') {
+              on('dev-server:start', (options) => devServer.devServer(options, result.component?.devServerConfig))
+
+              return setupNodeEvents(on, config)
             }
 
-            const setupNodeEvents = result.component?.setupNodeEvents ?? ((on, config) => {})
+            if (typeof devServer === 'function') {
+              on('dev-server:start', (options) => devServer(options, result.component?.devServerConfig))
 
-            return setupNodeEvents(on, config)
+              return setupNodeEvents(on, config)
+            }
+
+            if (typeof result.component.devServer.then === 'function') {
+              return Promise.resolve(result.component?.devServer).then(({ devServer }) => {
+                if (typeof devServer === 'function') {
+                  on('dev-server:start', (options) => devServer(options, result.component?.devServerConfig))
+                }
+
+                return setupNodeEvents(on, config)
+              })
+            }
```

Comment:
Completely missed all of this logic. We can probably clean it up a bit, but for now let's just get it working properly